File: 杰理之家/【杰理之家】SDK_v1.6.0】/NewJieliZhiNeng/NewJieliZhiNeng/App设置/关于/AppAboutVC.m
//
//  AppAboutVC.m
//  NewJieliZhiNeng
//
//  Created by kaka on 2020/5/15.
//  Copyright © 2020 杰理科技. All rights reserved.
//

#import "AppAboutVC.h"
#import "UserProfileVC.h"
#import "PrivacyPolicyVC.h"

@interface AppAboutVC ()<UITableViewDelegate,UITableViewDataSource>{
    UIImageView *topImv;
    __weak IBOutlet UIButton *backBtn;
    __weak IBOutlet UILabel *titleName;
    __weak IBOutlet UIView *subTitleView;
    
    UIView *currentVerUIView; //当前版本View
    UITableView *aboutTableView; //列表TableView
    NSMutableArray *itemArray;
    __weak IBOutlet NSLayoutConstraint *headHeight;
}

@end

@implementation AppAboutVC

- (void)viewDidLoad {
    [super viewDidLoad];
    [self addNote];
    [self initUI];
}

-(void)initUI{
    
    self.view.backgroundColor = kDF_RGBA(248, 250, 252, 1.0);
    
    float sw = [UIScreen mainScreen].bounds.size.width;
    subTitleView.frame = CGRectMake(0, 0, sw, kJL_HeightStatusBar+44);
    backBtn.frame  = CGRectMake(4, kJL_HeightStatusBar, 44, 44);
    titleName.text = kJL_TXT("about");
    titleName.bounds = CGRectMake(0, 0, 200, 20);
    titleName.center = CGPointMake(sw/2.0, kJL_HeightStatusBar+20);
    
    
    //名称
//    NSMutableAttributedString *titleNameStr = [[NSMutableAttributedString alloc] initWithString:kJL_TXT("about") attributes:@{NSFontAttributeName: [UIFont fontWithName:@"PingFang SC" size: 18],NSForegroundColorAttributeName: [UIColor colorWithRed:36/255.0 green:36/255.0 blue:36/255.0 alpha:1.0]}];
//    titleName.attributedText = titleNameStr;
    
    //顶部图标
    topImv = [[UIImageView alloc] initWithFrame:CGRectMake([UIScreen mainScreen].bounds.size.width/2-118/2,180,118,118)];
    topImv.image = [UIImage imageNamed:@"Theme.bundle/logo_1"];
    topImv.contentMode = UIViewContentModeScaleAspectFill;
    [self.view addSubview:topImv];
    
    aboutTableView = [[UITableView alloc] initWithFrame:CGRectMake(0,topImv.frame.origin.y+topImv.frame.size.height+80,[UIScreen mainScreen].bounds.size.width,55*3)];
    aboutTableView.delegate      = self;
    aboutTableView.dataSource    = self;
    aboutTableView.scrollEnabled = NO;
    aboutTableView.tag = 0 ;
    aboutTableView.rowHeight = 55;
    [aboutTableView setSeparatorColor:[UIColor colorWithRed:238/255.0 green:238/255.0 blue:238/255.0 alpha:1.0]];
    [self.view addSubview:aboutTableView];
    
    NSArray *tmpArray = @[kJL_TXT("firmware_current_version"),kJL_TXT("user_agreement"),kJL_TXT("privacy_policy_name")];
    
    [itemArray removeAllObjects];
    itemArray = nil;
    
    itemArray = [NSMutableArray arrayWithArray:tmpArray];
    [aboutTableView reloadData];
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return itemArray.count;
}


-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    static NSString* IDCell = @"lcell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:IDCell];
    if (cell == nil ) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:IDCell];
    }

    [cell.contentView setBackgroundColor:[UIColor colorWithRed:255/255.0 green:255/255.0 blue:255/255.0 alpha:1.0]];

    cell.textLabel.text = itemArray[indexPath.row];
    cell.textLabel.font = [UIFont fontWithName:@"PingFang SC" size: 14];
    cell.textLabel.textColor = kDF_RGBA(36, 36, 36, 1);
    
    if (@available(iOS 12.0, *)) {
        if(self.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark){
            cell.textLabel.alpha  = 0.7;
        }else{
            cell.textLabel.alpha  = 1.0;
        }
    } else {
        // Fallback on earlier versions
        cell.textLabel.alpha  = 1.0;
    }
    
    if(indexPath.row == 0){
        UILabel *currentVerCodeLabel = [[UILabel alloc] init];
        currentVerCodeLabel.frame = CGRectMake(self.view.frame.size.width-14.5-32,22.5,38,11);
        currentVerCodeLabel.numberOfLines = 0;
        [cell addSubview:currentVerCodeLabel];
        
        NSDictionary *infoDic = [[NSBundle mainBundle] infoDictionary];
        NSString *appVersion = [infoDic objectForKey:@"CFBundleShortVersionString"];
        NSMutableAttributedString *currentVerCodeStr = [[NSMutableAttributedString alloc] initWithString:appVersion attributes:@{NSFontAttributeName: [UIFont fontWithName:@"PingFang SC" size: 14],NSForegroundColorAttributeName: [UIColor colorWithRed:152/255.0 green:152/255.0 blue:152/255.0 alpha:1.0]}];
        
        currentVerCodeLabel.attributedText = currentVerCodeStr;
    }else{
        UIButton *nextBtn = [[UIButton alloc] initWithFrame:CGRectMake(self.view.frame.size.width-32,22.5,24.5,11)];
        [nextBtn setImage:[UIImage imageNamed:@"Theme.bundle/icon_app_settings_next"] forState:UIControlStateNormal];
        [cell addSubview:nextBtn];
    }
    return cell;
}



-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    if(indexPath.row == 1){
        UserProfileVC *vc = [[UserProfileVC alloc] init];
        vc.modalPresentationStyle = UIModalPresentationFullScreen;
        [self presentViewController:vc animated:YES completion:nil];
    }
    if(indexPath.row == 2){
        PrivacyPolicyVC *vc = [[PrivacyPolicyVC alloc] init];
        vc.modalPresentationStyle = UIModalPresentationFullScreen;
        [self presentViewController:vc animated:YES completion:nil];
    }
}

#pragma mark 退出关于界面
- (IBAction)backExit:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

-(void)noteDeviceChange:(NSNotification*)note{
    JLUuidType type = [note.object intValue];
    if (type == JLDeviceChangeTypeInUseOffline) {
        [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
    }
}

-(void)addNote{
    [JL_Tools add:kUI_JL_DEVICE_CHANGE Action:@selector(noteDeviceChange:) Own:self];
}
     
-(void)dealloc{
    [JL_Tools remove:nil Own:self];
}

@end

File: code/JieLi_Home_Demo/NewJieliZhiNeng/GlobalViews/AppSettingVC/AboutVC/AppAboutVC.swift
//
//  AppAboutVC.swift
//  NewJieliZhiNeng
//
//  Created by EzioChan on 2026/3/30.
//  Copyright © 2026 杰理科技. All rights reserved.
//

import UIKit
import SnapKit

@objcMembers
class AppAboutVC: BaseViewController, UITableViewDelegate, UITableViewDataSource {
    
    private var topImv: UIImageView!
    private var backBtn: UIButton!
    private var titleName: UILabel!
    private var subTitleView: UIView!
    
    private var currentVerUIView: UIView!
    private var aboutTableView: UITableView!
    private var itemArray: [String] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        addNote()
    }
    
    override func initUI() {
        super.initUI()
        
        self.navigationView.titleLab.text = R.Language.lan("about")
        self.view.backgroundColor = UIColor(red: 248/255.0, green: 250/255.0, blue: 252/255.0, alpha: 1.0)
        
        // 顶部图标
        topImv = UIImageView()
        topImv.image = UIImage(named: "Theme.bundle/logo_1")
        topImv.contentMode = .scaleAspectFill
        self.view.addSubview(topImv)
        
        topImv.snp.makeConstraints { make in
            make.top.equalTo(self.view).offset(180)
            make.centerX.equalTo(self.view)
            make.width.height.equalTo(118)
        }
        
        let gapes = UILongPressGestureRecognizer(target: self, action: #selector(longPressAction))
        gapes.minimumPressDuration = 4
        self.view.addGestureRecognizer(gapes)
        
        aboutTableView = UITableView()
        aboutTableView.delegate = self
        aboutTableView.dataSource = self
        aboutTableView.isScrollEnabled = false
        aboutTableView.tag = 0
        aboutTableView.rowHeight = 55
        aboutTableView.separatorColor = UIColor(red: 238/255.0, green: 238/255.0, blue: 238/255.0, alpha: 1.0)
        self.view.addSubview(aboutTableView)
        
        aboutTableView.snp.makeConstraints { make in
            make.top.equalTo(topImv.snp.bottom).offset(80)
            make.left.right.equalTo(self.view)
            make.height.equalTo(55 * 3)
        }
        
        let tmpArray = [R.Language.lan("firmware_current_version"), R.Language.lan("user_agreement"), R.Language.lan("privacy_policy_name")]
        
        itemArray.removeAll()
        itemArray.append(contentsOf: tmpArray)
        aboutTableView.reloadData()
    }
    
    // MARK: - long press
    @objc func longPressAction() {
        let vc = DebugSettingVC()
        self.navigationController?.pushViewController(vc, animated: true)
    }
    
    // MARK: - tableview delegate
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return itemArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let IDCell = "lcell"
        var cell = tableView.dequeueReusableCell(withIdentifier: IDCell)
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: IDCell)
        }
        
        cell?.contentView.backgroundColor = UIColor.white
        
        cell?.textLabel?.text = itemArray[indexPath.row]
        cell?.textLabel?.font = UIFont(name: "PingFang SC", size: 14)
        cell?.textLabel?.textColor = UIColor(red: 36/255.0, green: 36/255.0, blue: 36/255.0, alpha: 1.0)
        
        if #available(iOS 12.0, *) {
            if self.traitCollection.userInterfaceStyle == .dark {
                cell?.textLabel?.alpha = 0.7
            } else {
                cell?.textLabel?.alpha = 1.0
            }
        } else {
            cell?.textLabel?.alpha = 1.0
        }
        
        if indexPath.row == 0 {
            let currentVerCodeLabel = UILabel()
            currentVerCodeLabel.numberOfLines = 0
            cell?.contentView.addSubview(currentVerCodeLabel)
            
            currentVerCodeLabel.snp.makeConstraints { make in
                make.right.equalTo(cell!.contentView).offset(-14.5)
                make.centerY.equalTo(cell!.contentView)
            }
            
            if let infoDic = Bundle.main.infoDictionary, let appVersion = infoDic["CFBundleShortVersionString"] as? String {
                let attributes: [NSAttributedString.Key: Any] = [
                    .font: UIFont(name: "PingFang SC", size: 14) ?? UIFont.systemFont(ofSize: 14),
                    .foregroundColor: UIColor(red: 152/255.0, green: 152/255.0, blue: 152/255.0, alpha: 1.0)
                ]
                let currentVerCodeStr = NSAttributedString(string: appVersion, attributes: attributes)
                currentVerCodeLabel.attributedText = currentVerCodeStr
            }
        } else {
            let nextBtn = UIButton()
            nextBtn.setImage(UIImage(named: "Theme.bundle/icon_app_settings_next"), for: .normal)
            cell?.contentView.addSubview(nextBtn)
            
            nextBtn.snp.makeConstraints { make in
                make.right.equalTo(cell!.contentView).offset(-14.5)
                make.centerY.equalTo(cell!.contentView)
                make.width.equalTo(24.5)
                make.height.equalTo(11)
            }
        }
        return cell!
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        if indexPath.row == 1 {
            let vc = UserProfileVC()
            self.navigationController?.pushViewController(vc, animated: true)
        }
        if indexPath.row == 2 {
            let vc = PrivacyPolicyVC()
            self.navigationController?.pushViewController(vc, animated: true)
        }
    }
    
    @objc func noteDeviceChange(_ note: Notification) {
        if let typeValue = note.object as? Int, typeValue == 1 { // JLDeviceChangeTypeInUseOffline = 1
            self.navigationController?.popToRootViewController(animated: true)
        }
    }
    
    func addNote() {
        JL_Tools.add(kUI_JL_DEVICE_CHANGE, action: #selector(noteDeviceChange(_:)), own: self)
    }
    
    deinit {
        JL_Tools.remove(nil, own: self)
    }
}
